README.md
## Beskrivelse

Frontend-applikasjon for visning av stillinger på [arbeidsplassen.nav.no](https://arbeidsplassen.nav.no).

Applikasjonen har ansvar for søk og filtrering av stillinger, lagring av favorittstillinger og lagring av søk slik at
brukere kan få daglige oppdateringer pr. e-post.

Applikasjonen henter stillinger fra en dokumentdatabase (ElasticSearch) gjennom
[pam-search-api](https://github.com/navikt/pam-search-api). Lagrede søk og stillingsfavoritter, samt utsending av
e-poster skjer gjennom applikasjonen [pam-aduser](https://github.com/navikt/pam-aduser).

## Avhengigheter

-   [pam-search-api](https://github.com/navikt/pam-search-api)

# Teknisk dokumentasjon

## Teknologier

-   React
-   Node

## Systemlandskap

Bildet viser en forenklet skisse av pam-stillingsok og nærmeste integrasjoner.

![Teknisk skisse](images/teknisk-skisse.png)

### Frontend

Appens frontend er skrevet i React. Den viser stillinger, favoritter og lagrede søk. Brukere kan søke etter
stillinger uten å logge inn, mens favoritter og lagrede søk krever innlogging.

### Backend

Appen har en backend for frontend skrevet i JavaScript (Node). Backenden står for en del logikk, blant annet
konvertering av søkekriterier i frontend til ElasticSearch for å kunne utføre spørringer mot pam-search-api.

### Stillingsdatabase (ElasticSearch) og pam-search-api

[navikt/pam-search-api](http://github.com/navikt/pam-search-api) har en dokumentdatabase med stillinger
(ElasticSearch) som pam-stillingsok henter stillinger fra via REST.

En index-tjeneste henter stillinger fra stillingsdatabasen og indekserer dem til ElasticSearch via  
REST.

### Lagrede favoritter, lagrede søk og pam-aduser

[navikt/pam-aduser](http://github.com/navikt/pam-aduser) har funksjonalitet for lagring av
favorittstillinger, lagrede søk og utsending av epost med lagrede søk. Appen har et REST API som pam-stillingsok sin frontend bruker for å
hente og editere favoritter og lagrede søk.

Favorittstillinger lagres i en Postgres-database i pam-aduser. Favorittene synces mot
stillingsdatabasen med masterdata for stillinger via REST for å fange opp endringer i stillingannonsers status, tittel
osv.

Lagrede søk fungerer ved at pam-stillingsok genererer en predefinert spørring som kan eksekveres mot pam-search-api.
Denne spørringen lagres i pam-aduser. Hver natt kjøres alle lagrede spørringer mot pam-stillingsok. Nye
stillinger sendes til brukere over epost med Microsoft Graph API.

# Komme i gang

## Før kjøring av applikasjonen

Før du starter må du installere alle npm pakkene, dette kan du gjøre ved å kjøre kommandoen:

```
$ npm install
```

### Hvordan få tilgang til @navikt/arbeidsplassen-react og @navikt/arbeidsplassen-css

Opprett fila `.npmrc` i hjemkatalogen din. F.eks. `~/.npmrc` Mer info: https://docs.npmjs.com/cli/v9/configuring-npm/npmrc

Legg til følgende i fila

```
@navikt:registry=https://npm.pkg.github.com
//npm.pkg.github.com/:_authToken=$TOKEN
```

Opprett et token med "read:packages" rettigheter. [https://github.com/settings/tokens](https://github.com/settings/tokens) Bytt ut \$TOKEN med tokenet du akkurat opprettet. Velg Authorize token under "Configure SSO" for å gi tokenet tilgang til @navikt.

Ikke sjekk inn `.npmrc` til GitHub.

Mer informasjon om autentisering: https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-npm-registry#authenticating-with-a-personal-access-token

For å starte docker-containere for redis og mock-oauth2-server.

```
$ npm run start:dependencies
```

#### Go

Go er nødvendig for å bygge wonderwall binaries. Wonderwall brukes for lokal OIDC-flyt.

```
brew install go
```

## Kjøre applikasjonen lokalt

### Wonderwall

[Wonderwall](https://github.com/nais/wonderwall) brukes for å håndtere login med ID-porten.
Lokalt brukes også wonderwall, men her kjører vi mot en OIDC-provider i Docker.

Installer Wonderwall:

```
make install
```

Når applikasjonen er oppe, så kan du gå inn på [http://localhost:3000/stillinger](http://localhost:3000/stillinger)
Gå igjennom login-flyten ved å trykke login. Bruk testbruker `04010100653`

### Med dev script

Du kan enkelt kjøre applikasjonen ved hjelp av dev scriptet `./runDev.sh`
Scriptet starter opp wonderwall og setter følgende verdier som miljøvariabler:

```
PAMSEARCHAPI_URL=http://localhost:9000
PAMADUSER_URL=http://localhost:9017
PAM_STILLINGSOK_URL=http://localhost:8080/stillinger
PAM_VAR_SIDE_URL=http://localhost:8080/stillingsregistrering
PAM_INTERNAL_SEARCH_API_URL=http://localhost:9027
PAM_JOBBTREFF_API_URL=http://localhost:8088/jobbtreff-promo
AMPLITUDE_TOKEN=(Dev Token)
```

Applikasjonen vil da forsøke å koble seg til pam-search-api gjennom `localhost:9000`.
Dersom du vil få inn teststillinger kan du koble deg direkte til et kjørende instans av pam-search-api i kubernetes. Dette gjør du med følgende kommando:

```
$ kubectl config use-context dev-gcp
$ kubectl -n teampam port-forward deployment/pam-search-api 9000:9000
```

Dersom du får får feilmelding ved kjøring av kommandoene over kan du prøve å logge inn i Gcloud med følgende kommando.

```
gcloud auth login
```

Om du får følgende output betyr det at port-forwarden funket og pam-search-api er tilgjengelig på port 9000.

```
Forwarding from 127.0.0.1:9000 -> 9000
Forwarding from [::1]:9000 -> 9000
```

## Bruk av innloggede tjenester

For å kunne bruke innloggede tjenester (dvs. favoritter og lagrede søk), må du først kjøre `pam-aduser`.


releasenotes.md
## RELEASE - 0.210.1139-41e20b7

#### New in this release:

-   2020-05-25 [Fix] Css på intern stilling
-   2020-04-29 [Fix] Legger til overflow-wrap på tekst over stillingstittel

## RELEASE - 0.209.1137-4e6db67

#### New in this release:

-   2020-04-29 [Fix] Gjør ad text litt smalere på store skjemer

## RELEASE - 0.207.1134-3b16192

#### New in this release:

-   2020-04-29 [Fix] Dropper arbeidsplassen header fra print
-   2020-04-29 [Fix] Gjør adtext smalere
-   2020-04-29 [Fix] Gjør høyre kolonne bredere
-   2020-04-29 [Fix] Legger til luft i bunn av stillingsannonsen
-   2020-04-29 [Fix] Fikser plassering av tilbakelenke
-   2020-04-29 [Fix] Legger header i column
-   2020-04-29 [Fix] Ruller tilbake til grid, i istedet for flex
-   2020-04-29 [Fix] Ruller tilbake til grid, i istedet for flex
-   2020-04-29 [Fix] Css tweak for å unngå skaleringsproblem av mobilside ved veldig lange ord i stillingstittel
-   2020-04-29 [Fix] Css tweak for å unngå skaleringsproblem av mobilside ved veldig lange ord i stillingstittel - revert

## RELEASE - 0.206.1117-ab8ce0c

#### New in this release:

-   2020-04-22 [Fix] Css tweak for å unngå skaleringsproblem av mobilside ved veldig lange ord i stillingstittel - revert
-   2020-04-22 [Fix] Css tweak for å unngå skaleringsproblem av mobilside ved veldig lange ord i stillingstittel

## RELEASE - 0.205.1111-04d40db

#### New in this release:

-   2020-04-06 Revert "[Snyk] Upgrade nav-frontend-skjema from 1.0.60 to 1.0.78"
-   2020-04-06 Revert "[Snyk] Upgrade nav-frontend-skjema-style from 1.0.38 to 1.0.48"
-   2020-04-06 Revert "[Snyk] Upgrade nav-frontend-ekspanderbartpanel from 1.0.25 to 1.0.35"
-   2020-03-17 Bump acorn from 6.1.1 to 6.4.1

## RELEASE - 0.204.1106-73542b6

#### New in this release:

-   2020-04-04 [Fix] Wrap ga i try catch

## RELEASE - 0.197.1089-f381c4e

#### New in this release:

-   2020-03-24 [Fix] Fikser callback etter login også for stillingssiden

## RELEASE - 0.195.1081-60c8b72

#### New in this release:

-   2020-03-20 [FEATURE] supports lookup to internal index

## RELEASE - 0.189.1064-1a30f1a

#### New in this release:

-   2020-03-17 [FIx] Legger til menypunkter css tweak
-   2020-03-17 [FIx] Legger til menypunkter css tweak
-   2020-03-17 [FIx] Legger til menypunkter css tweak
-   2020-03-17 [FIx] Legger til menypunkter css tweak

## RELEASE - 0.188.1058-c284b25

#### New in this release:

-   2020-03-17 [FIx] Legger til menypunkter

## RELEASE - 0.175.1013-e36faa4

#### New in this release:

-   2020-01-22 [Fix] Fjerner hotjar feedbacklenke

## RELEASE - 0.172.1003-ed10be1

#### New in this release:

-   2020-01-21 [Fix] Fikser feil case i fylkesnavn
-   2020-01-21 [Fix] Fikser feil case i fylkesnavn

## RELEASE - 0.171.999-8c43186

#### New in this release:

-   2020-01-21 [Fix] Disabler qa-test
-   2020-01-21 [Fix] Disabler qa-test
-   2020-01-21 [Fix] Disabler qa-test
-   2020-01-21 [Fix] Disabler qa-test
-   2020-01-21 [Fix] Disabler qa-test
-   2020-01-21 [Fix] Disabler qa-test
-   2020-01-21 [Fix] Fjerner lenke-test for åpning av finn-annonser. Åpner alle finn-annonser i samme fane.

## RELEASE - 0.170.982-d43da32

#### New in this release:

-   2019-12-19 [PAM-3864] Legge til alertstripe med info om fylke/kommunesammenslåing

## RELEASE - 0.168.976-da72e5c

#### New in this release:

-   2019-12-17 [Fix] Legger til flere CSP headere

## RELEASE - 0.167.974-bcfa9b7

#### New in this release:

-   2019-12-17 [Fix] Legger til referrer policy header

## RELEASE - 0.166.973-23951aa

#### New in this release:

-   2019-12-17 [Fix] Legger til baseUri i CSP-headers

## RELEASE - 0.165.972-3d1900c

#### New in this release:

-   2019-12-17 [Fix] Husker hvilke av panelene som er åpne, slik de forblir åpnet når bruker kommer tilbake fra en ekstern side eller refresher nettsiden.
-   2019-12-17 [Fix] Unngår at QA-test velger ekstern annonse
-   2019-12-17 [Fix] Unngår at QA-test velger ekstern annonse
-   2019-12-17 [Fix] Legger inn igjen QA-tester
-   2019-08-29 [Fix] Finetuning av CSP og sikkerhetsheadere

## RELEASE - 0.164.960-6691698

#### New in this release:

-   2019-12-10 [Fix] Unngår dobbelt søk første gang
-   2019-12-06 [Fix] Oppdaterer dependecier

## RELEASE - 0.163.956-1e37e65

#### New in this release:

-   2019-12-05 [Fix] IE issue med isValidUrl

## RELEASE - 0.162.955-1ff7e75

#### New in this release:

-   2019-12-04 [Fix] Legger til lenke for eksterne annonser i annonsevisningen
-   2019-12-04 [Fix] Legger til lenke for eksterne annonser i annonsevisningen
-   2019-12-04 [Fix] Legger til lenke for eksterne annonser i annonsevisningen
-   2019-12-04 [Fix] Legger til lenke for eksterne annonser i annonsevisningen
-   2019-12-04 [Fix] Legger til lenke for eksterne annonser i annonsevisningen
-   2019-12-04 [Fix] Disabler QA-tester
-   2019-12-04 [Test] Tester ulike måter å åpne eksterne annonser
-   2019-12-04 [Test] Tester ulike måter å åpne eksterne annonser

## RELEASE - 0.161.939-d541f6f

#### New in this release:

-   2019-11-27 [Fix] Tester ny meta title
-   2019-11-27 [Fix] Tester ny meta title

## RELEASE - 0.160.936-b6a59b0

#### New in this release:

-   2019-11-26 [Fix] Legger til målinger for antall endringer av lagrede søk

## RELEASE - 0.159.935-c96042b

#### New in this release:

-   2019-11-20 [Fix] Legger til bilde som vises ved deling på facebook

## RELEASE - 0.155.927-3478c57

#### New in this release:

-   2019-11-13 [ES 7.2] peke på riktig search-api

## RELEASE - 0.154.926-b1a3120

#### New in this release:

-   2019-11-11 [ES 7.2] peker mot ny search-api

## RELEASE - 0.153.919-1f0862c

#### New in this release:

-   2019-10-09 [Bug] Fikser printknapp

## RELEASE - 0.152.913-9ca0a95

#### New in this release:

-   2019-10-09 [Fix] Legger til paginering i analytics

## RELEASE - 0.151.912-479d24f

#### New in this release:

-   2019-10-09 [Fix] Fjerner console.log
-   2019-10-09 [Fix] Rydder opp i google events

## RELEASE - 0.150.908-8ccc468

#### New in this release:


runDev.sh
#!/bin/bash
export PAMSEARCHAPI_URL=http://localhost:9000
export PAMADUSER_URL=http://localhost:9017/aduser
export PAM_STILLINGSOK_URL=http://localhost:3000/stillinger
export PAM_VAR_SIDE_URL=http://localhost:8080/stillingsregistrering
export PAM_INTERNAL_SEARCH_API_URL=http://localhost:9027
export PAM_JOBBTREFF_API_URL=http://localhost:8088/jobbtreff-promo
export AMPLITUDE_TOKEN=b6143dedf1d17c3e73d3ad440bf54e8b
export NAIS_CLUSTER_NAME=local-dev
export TOKEN_X_WELL_KNOWN_URL=http://host.docker.internal:8233/tokenx/.well-known/openid-configuration
export IDPORTEN_WELL_KNOWN_URL=http://host.docker.internal:8233/idporten/.well-known/openid-configuration
export TOKEN_X_CLIENT_ID=local-token-x-client-id
export TOKEN_X_ISSUER=http://host.docker.internal:8233/tokenx
export TOKEN_X_TOKEN_ENDPOINT=http://host.docker.internal:8233/tokenx/token
export IDPORTEN_JWKS_URI=http://host.docker.internal:8233/idporten/jwks
export TOKEN_X_PRIVATE_JWK='{"p":"1KMmKh85qonPCvMfbIyDbaSWeeBuMgJ3ISlLwV768PMR3vyFmaahYkAwRlG2QvDG4c4UDzqroaWlZ_5REC4M0oMFlWvZ6N4raWKteqyUxFChjDMpsFj1o_i4XYkPp_rB3ehMVvpVE64EylTe9_E0cLwFir82xA6pP8ho_BC6jtk","kty":"RSA","q":"nvKKaS1AG02qKnlFSYtKxyEXu-6sJr8GdSO0IYgvxBIboIxXwW14OqFY8aBQTDCrtXCSkwhQyCYwrhBZCpWgFhTedxefKfdzKzB9E4peaymnKOfjX4BaliG_j8z9_yavPOFDgsDdccbb5tqPq6OkCmZpTveIfygaJFc9N1-Fbr8","d":"JRdFU4dUucb8T6_9BR34B9x8zeyZA4vKN0hU2Ne5sf9_uahAsUAxSuQloAUC_I8yVURYC-ldl0RGNtH6FqWYZGaKo-92haOv6KRdHH4--8e44IIz3_3ySrTSskJyW018mP5uj6riCl-NJKCoAR4PfcwHXxj0fTZpdkm0g_ZuQeHyhbccgGO5y4xpFblzxmNv9FOwFtfsmRhdbp4XTtghFmT5y-eK5WhLA2evv_9zZRwgpp-BKjEqW2vBvEznnHDLGrK2pCtMrLv7VFKeaQlk01jRh-Old0Cola8xRHnPNFRVuGcaJbxbc83kJgPosK7gwMRU1ShOArPEIRcsWMzpwQ","e":"AQAB","use":"sig","kid":"jXDxKRE6a4jogcc4HgkDq3uVgQ0","qi":"k_FPIPmDVCwjxhdNb0zLg16bght6nkGfxPgS57dRslN_eEWy_FJNdDMmRQ17I0XLal-0ciRO4Q3pJg52vPX84y-tmA6OkTo5cbPX9NOhUz5I0bdahGLsBADI8Hod9g-WPw9peIaF9K1U335j_-M4Bw-35XL9kfFExrv28Ael594","dp":"h4LqAZIgyARHRG0oq57o62km1MkosQpNLX8J3lRPjFr8GqXH5uIAm4LYwRe2FfcoaCNTOVlkpAHS5fF3a95NIaOKXW18VCGyfNj8Qd_Wdz4DljvM56vp7_PEWyfMKHtwLaGrdQrYSVBvkRqBdnXru6S3y3KMzmmhhxqLfYiTvBE","alg":"RS256","dq":"fkdNRaBItShv0AzGylYa07dr5sgJB0NuuvPhsV5ELK8COfwl4Cu7gEEdq-HkyZNFaAIGwT-zBkCpHjEovKsBI2nzEIS41Bue6So9xJ-Cwfq2rl4Em6Qa0_n9-h4vkV3HjUOeYiweAr2tiWwNBp0cv5J7idxJGlTe6A5TOUHvfXs","n":"hAYmv2IqZQzSGTK7RixXgl7yEu3MgG2kuK4inAK0cSOseyIgDj_-X9ihkQb-aLMRW82vlTyypaZsYf9025EzI4YTnHoOMoerMLkrcqvKTZ8ujcCMLr4gfR-b_Wbl4oV6DkegiWQwIA6QjFgMe403nSnzXPBPgVjpUYV2Jr_n3fi5tjEwr07sX5ij9j6w3OW1aJMTToTmJUdlkS2S4ny9V5LVQD-5P9vGFpNUMaOcJw7lXiG1eQV8fcfrkUmN2jSa5A3kAniVsvrZGQ-QbzsC8IvZGVfiVvqv_G7KoJDYqJUBYZJVRAwbdHOStNsEMcmmyb2f8xsvH23kqfTOYk7R5w"}'
export ADUSER_AUDIENCE=local:teampam:pam-aduser
export IDPORTEN_AUDIENCE=local-token-x-client-id
export CV_API_AUDIENCE=local:teampam:pam-cv-api
export ARBEIDSPLASSEN_URL=http://localhost:3000

exec npm start
